File: frontend/src/types/compliance.ts
/**
 * 合规检查相关类型定义
 * Compliance check related TypeScript types
 */

// ─────────────────────────────────────────────
// 枚举字面量联合类型
// ─────────────────────────────────────────────

export type RuleType = 'number' | 'name' | 'description' | 'file';

export type RuleSeverity = 'must' | 'should';

export type ComplianceCheckStatus = 'pending' | 'completed' | 'failed';

// ─────────────────────────────────────────────
// 规则集合 (Rule Set)
// ─────────────────────────────────────────────

export interface RuleSet {
  id: string;
  name: string;
  description: string | null;
  is_active: boolean;
  rule_count: number;
  created_at: string;
  updated_at: string;
}

export interface RuleSetDetail extends RuleSet {
  rules: Rule[];
}

export interface CreateRuleSetDto {
  name: string;
  description?: string;
  is_active?: boolean;
}

export interface UpdateRuleSetDto {
  name?: string;
  description?: string;
  is_active?: boolean;
}

// ─────────────────────────────────────────────
// 规则 (Rule)
// ─────────────────────────────────────────────

export interface Rule {
  id: string;
  rule_set_id: string;
  rule_type: RuleType;
  title: string;
  requirement: string;
  severity: RuleSeverity;
  order: number;
  created_at: string;
  updated_at: string;
}

export interface CreateRuleDto {
  rule_set_id: string;
  rule_type: RuleType;
  title: string;
  requirement: string;
  severity?: RuleSeverity;
  order?: number;
}

export interface UpdateRuleDto {
  rule_id: string;
  rule_type?: RuleType;
  title?: string;
  requirement?: string;
  severity?: RuleSeverity;
  order?: number;
}

// ─────────────────────────────────────────────
// 违规项 (Violation)
// ─────────────────────────────────────────────

export interface Violation {
  rule_id: string;
  rule_title: string;
  rule_type: RuleType;
  location: RuleType;
  excerpt: string;
  description: string;
  suggestion: string;
  severity: RuleSeverity;
}

// ─────────────────────────────────────────────
// 合规检查结果 (Check Result)
// ─────────────────────────────────────────────

export interface UserBrief {
  id: string;
  name: string;
  avatar: string | null;
}

export interface ComplianceCheckResult {
  id: string;
  status: ComplianceCheckStatus;
  requested_by: UserBrief;
  rule_set_id: string | null;
  rule_set_name: string | null;
  file_name: string;
  file_size: number;
  file_mime_type: string;
  extracted_text: string;
  text_truncated: boolean;
  number_draft: string | null;
  name_draft: string | null;
  description_draft: string | null;
  violations: Violation[];
  suggested_name: string | null;
  suggested_description: string | null;
  compliance_score: number | null; // 0..100, status !== 'completed' 时为 null
  requested_at: string;
  completed_at: string | null;
  error_message: string | null;
  // 注意: 不包含 suggested_number / contract_text 字段
}

export interface ComplianceCheckSummary {
  id: string;
  status: ComplianceCheckStatus;
  name_draft: string | null;
  rule_set_name: string | null;
  file_name: string;
  text_truncated: boolean;
  violation_count: number | null; // 仅 completed 才有
  compliance_score: number | null; // 0..100, status !== 'completed' 时为 null
  requested_at: string;
  completed_at: string | null;
}

export interface ComplianceCheckListResponse {
  items: ComplianceCheckSummary[];
  total: number;
  page: number;
  page_size: number;
}

// ─────────────────────────────────────────────
// 查询参数
// ─────────────────────────────────────────────

export interface ListChecksParams {
  page?: number;
  page_size?: number;
  status?: ComplianceCheckStatus;
}

// ─────────────────────────────────────────────
// Excel 批量导入相关类型
// ─────────────────────────────────────────────

/** 预览接口返回的单条规则（含 Excel 行号） */
export interface ImportPreviewRule {
  row_number: number;
  rule_type: RuleType;
  title: string;
  requirement: string;
  severity: RuleSeverity;
  order: number;
}

/** 预览接口响应 */
export interface ImportPreviewResponse {
  preview_session_token: string;
  rules: ImportPreviewRule[];
  total_count: number;
}

/** 确认导入接口响应 */
export interface ImportConfirmResponse {
  imported_count: number;
  rule_set_id: string;
}

/** 行级校验错误 */
export interface ImportRowError {
  row_number: number;
  field: string;
  message: string;
}

/** 422 校验失败响应体 detail */
export interface ImportValidationError {
  code: string;
  message: string;
  errors?: ImportRowError[];
  current_count?: number;
  import_count?: number;
  limit?: number;
}

File: backend/app/schemas/compliance.py
"""
合规检查相关 Pydantic schemas
Compliance check related Pydantic schemas
"""

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field, field_validator


# ─────────────────────────────────────────────
# 规则集合 (Rule Set)
# ─────────────────────────────────────────────


class CreateRuleSetRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=100, description="规则集合名称")
    description: Optional[str] = Field(
        None, max_length=1000, description="规则集合描述"
    )
    is_active: bool = Field(False, description="是否为当前生效的规则集合")

    @field_validator("name")
    @classmethod
    def name_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("name 不能为空白字符串")
        return v


class UpdateRuleSetRequest(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=1000)
    is_active: Optional[bool] = None

    @field_validator("name")
    @classmethod
    def name_not_blank(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and not v.strip():
            raise ValueError("name 不能为空白字符串")
        return v


class RuleSetResponse(BaseModel):
    id: str
    name: str
    description: Optional[str]
    is_active: bool
    rule_count: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ─────────────────────────────────────────────
# 规则 (Rule)
# ─────────────────────────────────────────────


class CreateRuleRequest(BaseModel):
    rule_type: Literal["number", "name", "description", "file"] = Field(
        ..., description="规则作用对象"
    )
    title: str = Field(..., min_length=1, max_length=100, description="规则名称")
    requirement: str = Field(
        ..., min_length=1, max_length=2000, description="规则正文描述"
    )
    severity: Literal["must", "should"] = Field("must", description="严重程度")
    order: int = Field(0, description="排序权重")

    @field_validator("title")
    @classmethod
    def title_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("title 不能为空白字符串")
        return v

    @field_validator("requirement")
    @classmethod
    def requirement_not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("requirement 不能为空白字符串")
        return v


class UpdateRuleRequest(BaseModel):
    rule_type: Optional[Literal["number", "name", "description", "file"]] = None
    title: Optional[str] = Field(None, min_length=1, max_length=100)
    requirement: Optional[str] = Field(None, min_length=1, max_length=2000)
    severity: Optional[Literal["must", "should"]] = None
    order: Optional[int] = None

    @field_validator("title")
    @classmethod
    def title_not_blank(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and not v.strip():
            raise ValueError("title 不能为空白字符串")
        return v

    @field_validator("requirement")
    @classmethod
    def requirement_not_blank(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and not v.strip():
            raise ValueError("requirement 不能为空白字符串")
        return v


class RuleResponse(BaseModel):
    id: str
    rule_set_id: str
    rule_type: Literal["number", "name", "description", "file"]
    title: str
    requirement: str
    severity: Literal["must", "should"]
    order: int
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ─────────────────────────────────────────────
# 合规检查结果 (Check Result)
# ─────────────────────────────────────────────


class UserBriefDto(BaseModel):
    id: str
    name: str
    avatar: Optional[str]

    model_config = {"from_attributes": True}


class ViolationDto(BaseModel):
    rule_id: str
    rule_title: str  # 后端 join 补全
    rule_type: Literal["number", "name", "description", "file"]
    location: Literal["number", "name", "description", "file"]
    excerpt: str  # ≤ 500
    description: str  # ≤ 500
    suggestion: str  # ≤ 500
    severity: Literal["must", "should"]


class ComplianceCheckResultDto(BaseModel):
    id: str
    status: Literal["pending", "completed", "failed"]
    requested_by: UserBriefDto
    rule_set_id: Optional[str]
    rule_set_name: Optional[str]
    file_name: str
    file_size: int
    file_mime_type: str
    extracted_text: str  # 0-100k
    text_truncated: bool
    number_draft: Optional[str]
    name_draft: Optional[str]
    description_draft: Optional[str]
    violations: list[ViolationDto]
    suggested_name: Optional[str]
    suggested_description: Optional[str]
    compliance_score: Optional[int]  # 0..100, status != completed 时为 None
    requested_at: datetime
    completed_at: Optional[datetime]
    error_message: Optional[str]
    # 注意: 不包含 suggested_number / contract_text 字段 (R5.1)

    model_config = {"from_attributes": True}


class ComplianceCheckListItemDto(BaseModel):
    id: str
    status: Literal["pending", "completed", "failed"]
    name_draft: Optional[str]
    rule_set_name: Optional[str]
    file_name: str
    text_truncated: bool
    violation_count: Optional[int]  # 仅 completed 才有，否则 None
    compliance_score: Optional[int]  # 0..100, status != completed 时为 None
    requested_at: datetime
    completed_at: Optional[datetime]
    # 不包含 contract_text / extracted_text (R5.2)

    model_config = {"from_attributes": True}


class ComplianceCheckListResponse(BaseModel):
    items: list[ComplianceCheckListItemDto]
    total: int
    page: int
    page_size: int
